### PR TITLE
[GEOT-6799] WMTS online tests

### PR DESCRIPTION
**Pull request by roarbra**
This PR tries to fix up in WMTS online tests. Two goals.

- Extend OnelineTestCase and place them in online package
- Split into classes for separate server environments

The first one is to clarify what is online tests, and what isn't. The other one is for running some of the tests, but not necessary all.

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md)
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [X] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [X] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [X] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [X] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [X] PR for bug fixes and small new features are presented as a single commit
- [X] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

File: modules/extension/wmts/src/test/java/org/geotools/ows/wmts/online/LantmaterietWMTSServerOnlineTest.java [hunk 1]
```diff
@@ -45,22 +40,27 @@
...
+public class LantmaterietWMTSServerOnlineTest extends OnlineTestCase {
 
     private URL server;
 
     private URL serverWithStyle;
 
-    @Before
-    public void setUpInternal() throws MalformedURLException {
+    @Override
+    protected String getFixtureId() {
```
Comment (aaime) [inline modules/extension/wmts/src/test/java/org/geotools/ows/wmts/online/LantmaterietWMTSServerOnlineTest.java:58]: So this one needs a property file, but it's not actually going to use it?

File: modules/extension/wmts/src/test/java/org/geotools/ows/wmts/online/WMTSTransparentTileOnlineTest.java [hunk 2]
```diff
@@ -36,30 +38,24 @@
...
-    protected void tearDownInternal() throws Exception {
-        service = null;
+    public void setUpInternal() throws Exception {
+        service = createKVPService();
     }
 
     @Override
     protected String getFixtureId() {
-        return "wmts";
+        return "wmts-unesco";
```
Comment (aaime) [inline modules/extension/wmts/src/test/java/org/geotools/ows/wmts/online/WMTSTransparentTileOnlineTest.java:58]: Guess this one does not really use the file either?

**Review by aaime — COMMENTED** [review]
It's probably best if @ianturton and/or @etj have a look at this one, just added a couple of comments.

One thing that I believe is missing, if we actually want the tests to be executed, is a github action that does exactly that ([here is a postgis one as an example](https://github.com/geotools/geotools/blob/master/.github/workflows/postgis_online.yml)). 

Otherwise nobody will actually run them, besides maybe when working on the module.
I guess that these tests will have some tendency to randomly fail, as they depend on external servers that we have no control over.

On Travis we had a way to mark a check as not mandatory... I have not used this before, but it seems there is a "continue-on-fail" setting that might help, described in this issue:
https://github.com/actions/toolkit/issues/399
From what I understand, there are issues if there is a build matrix, but it may work well for an action that has a single build action in it.

**Comment by roarbra** [comment]
My intention was to clean up the tests within gt-wmts. First off was to mark those tests using external servers as so, by extending OnlineTestCase as this seems to be the GeoTools way of doing it as for now.

Next thing I noticed was that if several tests use the same fixtureId, and one of them fails within setUpInternal. Then this function wan't be called for the rest of the tests. Which occasionally made some tests fail at random. The same tests would work fine if started separately. This makes it confusing to work with the test cases.

I see the problem with many property files, but I'm not tempted to change the internals of OnlineTestCase either. The whole matter could maybe be solved differently, and I haven't actually looked into what these tests are testing.

What I do know is that someone, if that is me, @ianturton or @etj , should look into the tests of gt-wmts.

**Comment by aaime** [comment]
@roarbra whether the PR is going to be merged, or not, it's not my call, [I'm not the maintainer nor a developer of the gt-wmts module.](https://github.com/geotools/geotools/blob/master/modules/extension/wmts/pom.xml#L48) 

I'll refrain from adding further comments.

**Comment by aaime** [comment]
Looks like @etj won't be able to review this one for a while, so having a second look and actually trying to run the tests.

The "lantmariet" case does not seem to require a specific url, but just the creation of the property file (an example file is not created either). Is that right?
The actual test run is failing for me (running them with -Ponline from maven):

```
[ERROR]   LantmaterietWMTSServerOnlineTest>OnlineTestCase.run:117->testWMTSCoverageReaderWithStylePlaceholder:193 Images are visibly different, found 520 different pixels, against a threshold of 100
You can add -Dorg.geotools.image.test.interactive=true to show a dialog comparing them (requires GUI support)
```

Comparing with master, it fails the same way there....

The WMTSRestTest example file assumes a "raspberry" URL in the example file (@ianturton is that your doing?), from the looks of the test code it would seem it's looking for a MapProxy deploy? I could be wrong.
I guess the rest case would require a capabilities document that's purely resource oriented... I don't have one, nor the inclination to set up one. Tried giving it the caps document of a embedded GWC, on master (not this branch):

```
[ERROR]   WebMapTileServerOnlineTest>OnlineTestCase.run:117->testIssueGetTileRequestKVP:113->issueGetTileRequest:142 » NullPointer
[ERROR]   WebMapTileServerOnlineTest>OnlineTestCase.run:117->testIssueGetTileRequestREST:119->issueGetTileRequest:142 » NullPointer
[ERROR]   WMTSTransparentTileTest>OnlineTestCase.run:117->testTransparentTile:100 NullPointer
```

Oh well, the online tests are pretty broken to start with, and does not seem this change is making things any worse.

**Pull request merged**